pkg/osquery/tables/macos_software_update/recommended_updates_table.go

```diff
@@ -0,0 +1,113 @@
+//go:build darwin
+// +build darwin
+
+package macos_software_update
+
+/*
+#cgo darwin CFLAGS: -DDARWIN -x objective-c
+#cgo darwin LDFLAGS: -framework Cocoa
+#import <Cocoa/Cocoa.h>
+#import <SUSharedPrefs.h>
+#include "sus.h"
+*/
+import "C"
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/launcher/pkg/dataflatten"
+	"github.com/kolide/launcher/pkg/osquery/tables/dataflattentable"
+	"github.com/kolide/launcher/pkg/osquery/tables/tablehelpers"
+	"github.com/osquery/osquery-go/plugin/table"
+)
+
+const softwareUpdateToolPath = "/usr/sbin/softwareupdate"
+const softwareUpdateListArg = "--list"
+const softwareUpdateNoScanArg = "--no-scan"
+
+var updatesData []map[string]interface{}
+var doNotScan bool
+
+type Table struct {
+	logger log.Logger
+}
+
+func RecommendedUpdates(logger log.Logger) *table.Plugin {
+	columns := dataflattentable.Columns(
+		table.TextColumn("noscan"),
+	)
+
+	tableName := "kolide_macos_recommended_updates"
+
+	t := &Table{
+		logger: log.With(logger, "table", tableName),
+	}
+
+	return table.NewPlugin(tableName, columns, t.generate)
+}
+
+func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	var results []map[string]string
+	var err error
+
+	for _, noscan := range tablehelpers.GetConstraints(queryContext, "noscan", tablehelpers.WithDefaults("false")) {
+		doNotScan, err = strconv.ParseBool(noscan)
+		if err != nil {
+			level.Info(t.logger).Log("msg", "Cannot convert noscan constraint into a boolean value. Try passing \"true\"", "err", err)
+			continue
+		}
+
+		softwareUpdateArgs := []string{softwareUpdateListArg}
+
+		if doNotScan {
+			softwareUpdateArgs = append(softwareUpdateArgs, softwareUpdateNoScanArg)
+		}
+
+		_, err = tablehelpers.Exec(ctx, t.logger, 30, []string{softwareUpdateToolPath}, softwareUpdateArgs)
+		if err != nil {
+			level.Info(t.logger).Log("msg", "softwareupdate failed", "err", err)
+			return nil, nil
+		}
+
+		data := getUpdates()
```

Comment:
I'm a little confused by this. It's doing an exec, tossing the output, then using the frameworks to pull the data? Why have both approaches?

Reply:
So I _thought_ there was some correlation between `softwareupdate --list` invocations and the shared prefs API, but I did some more testing and I can still get the shared prefs API to give me update data regardless if `softwareupdate` is called, the state of `/Library/Updates/` dir, or the state/presence of the SoftwareUpdate cached plist.

The plist will be updated after calling `softwareupdate --list` but that seems to be independent of the framework itself.

I also noticed that even if I disable all macOS software updates behaviors, the preferences for checking & downloading updates gets reset to enabled after reboot 🤷  So even after blowing away all on-disk cached update data, killing the software update daemons, and rebooting, querying the `[SUSharedPrefs recommendedUpdates]` still gives me good data.

This is all to say - I've removed the exec entirely as it seems to have no benefit after all.